opentelemetry-api/opentelemetry/backend.py

```diff
@@ -0,0 +1,131 @@
+# Copyright 2019, OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""
+The OpenTelemetry backend module defines and implements the API to access an
+implementation (backend) of the OpenTelemetry API.
+
+By default, if you call a getter function (e.g., :func:`tracer`) and the corresponding setter (e.g.,
+:func:`set_tracer`) wasn't called, you will get a default implementation, which is selected as follows:
+
+    1. If the environment variable OPENTELEMETRY_PYTHON_BACKEND_<getter-name>` (e.g.,
+       OPENTELEMETRY_PYTHON_BACKEND_TRACER) is set to an nonempty value, an attempt is made to
+       import a module with that name and call a function `get_opentelemetry_backend_impl` in it.
+       The function receives the API type that is expected as an argument and should return an
+       instance of it (e.g., the argument is :cls:`opentelemetry.trace.Tracer` and the function should
+       return an instance of a :cls:`~Tracer` (probably of a derived type).
+    2. If the variable is not set, `$OPENTELEMETRY_PYTHON_BACKEND_DEFAULT` is tried instead.
+    3. If that variable was also not set, an attempt is made to import and use the OpenTelemetry
+       SDK.
+    4. Otherwise (if no variable was set and the SDK was not importable, or an error occured when
+       trying to instantiate the implementation object) the default implementation that ships with the
+        API distribution (a fast no-op implementation) is used.
+
+If you called the setter or if you call the setter later (not recommended), the object you set is
+used instead.
+"""
+
+import sys
+import os
+import importlib
+from typing import TypeVar, Type, Callable
+from functools import wraps
+
+from .trace import Tracer
```

Comment:
Personally, I prefer the style I used here (except that I didn't sort the imports here). I'm OK with not importing classes/functions, but if importing classes/functions, I'm against using a separate statement per name. This would make for a Java-like experience where the import section is routinely so long that most tools are hiding it by default.

One thing to note is that importing the used names directly saves one dictionary lookup per access to the name (the dot operator is not free at runtime).

Reply:
We can hash out import style in #36.